pkg/core/statesync/module.go

```diff
@@ -235,42 +287,65 @@ func (s *Module) defineSyncStage() error {
 		}
 		var mode mpt.TrieMode
-		// No need to enable GC here, it only has latest things.
+		// No need to enable GC here, it only has the latest things.
 		if s.bc.GetConfig().Ledger.KeepOnlyLatestState || s.bc.GetConfig().Ledger.RemoveUntraceableBlocks {
 			mode |= mpt.ModeLatest
 		}
-		s.billet = mpt.NewBillet(header.PrevStateRoot, mode,
-			TemporaryPrefix(s.dao.Version.StoragePrefix), s.dao.Store)
-		s.log.Info("MPT billet initialized",
-			zap.Uint32("height", s.syncPoint),
-			zap.String("state root", header.PrevStateRoot.StringBE()))
-		pool := NewPool()
-		pool.Add(header.PrevStateRoot, []byte{})
-		err = s.billet.Traverse(func(_ []byte, n mpt.Node, _ []byte) bool {
-			nPaths, ok := pool.TryGet(n.Hash())
-			if !ok {
-				// if this situation occurs, then it's a bug in MPT pool or Traverse.
-				panic("failed to get MPT node from the pool")
+		if s.mode == MPTBased {
+			s.billet = mpt.NewBillet(header.PrevStateRoot, mode,
+				TemporaryPrefix(s.dao.Version.StoragePrefix), s.dao.Store)
+			s.log.Info("MPT billet initialized",
+				zap.Uint32("height", s.syncPoint),
+				zap.String("state root", header.PrevStateRoot.StringBE()))
+			pool := NewPool()
+			pool.Add(header.PrevStateRoot, []byte{})
+			err = s.billet.Traverse(func(_ []byte, n mpt.Node, _ []byte) bool {
+				nPaths, ok := pool.TryGet(n.Hash())
+				if !ok {
+					// if this situation occurs, then it's a bug in MPT pool or Traverse.
+					panic("failed to get MPT node from the pool")
+				}
+				pool.Remove(n.Hash())
+				childrenPaths := make(map[util.Uint256][][]byte)
+				for _, path := range nPaths {
+					nChildrenPaths := mpt.GetChildrenPaths(path, n)
+					for hash, paths := range nChildrenPaths {
+						childrenPaths[hash] = append(childrenPaths[hash], paths...)
+					}
+				}
+				pool.Update(nil, childrenPaths)
+				return false
+			}, true)
+			if err != nil {
+				return fmt.Errorf("failed to traverse MPT during initialization: %w", err)
 			}
-			pool.Remove(n.Hash())
-			childrenPaths := make(map[util.Uint256][][]byte)
-			for _, path := range nPaths {
-				nChildrenPaths := mpt.GetChildrenPaths(path, n)
-				for hash, paths := range nChildrenPaths {
-					childrenPaths[hash] = append(childrenPaths[hash], paths...) // it's OK to have duplicates, they'll be handled by mempool
+			s.mptpool.Update(nil, pool.GetAll())
+			if s.mptpool.Count() == 0 {
+				s.syncStage |= mptSynced
+				s.log.Info("MPT is in sync",
+					zap.Uint32("stateroot height", s.syncPoint))
+			}
+		} else {
+			var p checkpoint
+			data, err := s.dao.Store.Get([]byte{byte(storage.SYSStateSyncCheckpoint)})
+			if err == nil {
+				br := io.NewBinReaderFromBuf(data)
+				p.DecodeBinary(br)
+				if br.Err != nil {
+					return fmt.Errorf("failed to decode checkpoint metadata: %w", br.Err)
 				}
+				s.localTrie = mpt.NewTrie(mpt.NewHashNode(p.mptRoot), mode, s.dao.Store)
+				s.lastStoredKey = p.lastKey
+			} else {
+				s.localTrie = mpt.NewTrie(nil, mode, s.dao.Store)
+			}
```

Comment:
Here we need to distinguish `storage.KeyNotFound` (it's OK, init clean trie) from some side error (return this error).